File: ardrone_operator/src/ardrone_keyboard_operation.cpp
#include "ros/ros.h"
#include <keyboard/Key.h>
#include <std_msgs/Empty.h>
#include <geometry_msgs/Twist.h>

enum Key
{
    UP = 273,
    DOWN = 274,
    RIGHT = 275,
    LEFT = 276,
    ENTER = 13,
    SPACE = 32,
    BACKSPACE = 8,
    KEY_w = 119,
    KEY_s = 115,
    KEY_a = 97,
    KEY_d = 100,
};

class KeyboardOperation
{
public:
    KeyboardOperation();
    ~KeyboardOperation();
private:
    void callbackKeyDown(const keyboard::KeyConstPtr &key_msg);
    void callbackKeyUp(const keyboard::KeyConstPtr &key_msg);

    ros::NodeHandle n_;
    ros::Subscriber sub_keydown_, sub_keyup_;
    ros::Publisher pub_vel_;
    ros::Publisher pub_takeoff_;
    ros::Publisher pub_land_;
    ros::Publisher pub_reset_;

    geometry_msgs::Twist vel_;
    std_msgs::Empty empty_msg_;
}

KeyboardOperation::KeyboardOperation()
{
    sub_keydown_ = n_.subscribe("/keyboard/keydown",10);
    sub_keyup_ = n_.subscribe("/keyboard/keyup",10);
    pub_takeoff_ = n_.advertise<std_msgs::Empty>("/ardrone/takeoff",1);
    pub_land_ = n_.advertise<std_msgs::Empty>("/ardrone/land",1);
    pub_reset_ = n_.advertise<std_msgs::Empty>("/ardrone/reset",1);
    pub_vel_ = n_.advertise<geometry_msgs::Twist>("/cmd_vel", 10);
}

KeyboardOperation::~KeyboardOperation()
{
}

KeyboardOperation::callbackKeyDown(const keyboard::KeyConstPtr &key_msg)
{
    int key_code = key_msg->code;
    switch (key_code)
    {
        case Key::Enter:
            pub_takeoff_.publish(empty_msg_);
            break;
        case Key::SPACE:
            pub_land_.publish(empty_msg_);
            break;
        case Key::BACKSPACE:
            pub_reset_.publish(empty_msg_);
        case Key::UP:
            vel_.linear.z += 1.0;
            break;
        case Key::DOWN:
            vel_.linear.z -= 1.0;
            break;
        case Key::LEFT:
            vel_.angular.z += 1.0;
            break;
        case Key::RIGHT:
            vel_.angular.z -= 1.0;
            break;
        case Key::KEY_w:
            vel_.linear.x += 1.0;
            break;
        case Key::KEY_s:
            vel_.linear.x -= 1.0;
            break;
        case Key::KEY_a:
            vel_.linear.y += 1.0;
            break;
        case Key::KEY_d:
            vel_.linear.y -= 1.0;
            break;
    }
    pub_vel_.publish(vel_);
}

KeyboardOperation::callbackKeyUp(const keyboard::KeyConstPtr &key_msg)
{
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "ardrone_keyboard_operation");
    ros::Rate loop_rate(10);

    KeyboardOperation key;

    ros::spin();
    return 0;
}

File: ardrone_operator/script/ardrone_keyboard_operation.py
#!/usr/bin/env python
# coding: UTF-8

import rospy
from std_msgs.msg import Empty
from geometry_msgs.msg import Twist
from keyboard.msg import Key

UP = 273
DOWN = 274
RIGHT = 275
LEFT = 276
ENTER = 13
SPACE = 32
BACKSPACE = 8
KEY_w = 119
KEY_s = 115
KEY_a = 97
KEY_d = 100

VELOCITY_INCLIMENT = 0.5

class KeyboardOperation:

    empty_msg = Empty()
    ardrone_vel = Twist()
    ardrone_vel.linear.x = 0
    ardrone_vel.linear.y = 0
    ardrone_vel.linear.z = 0
    ardrone_vel.angular.z = 0

    def __init__(self):
        rospy.init_node('ardrone_keyboard_operation')
        self.pub_takeoff = rospy.Publisher('ardrone/takeoff', Empty, queue_size=1)
        self.pub_land = rospy.Publisher('ardrone/land', Empty, queue_size=1)
        self.pub_reset = rospy.Publisher('ardrone/reset', Empty, queue_size=1)
        self.pub_vel = rospy.Publisher('cmd_vel', Twist, queue_size=10)

    def subscribe_key_data(self):
        rospy.Subscriber("keyboard/keydown", Key, self.keydownCallback)
        rospy.Subscriber("keyboard/keyup", Key, self.keyupCallback)

    def keydownCallback(self, key):
        if key.code == ENTER:
            self.pub_takeoff.publish(self.empty_msg)
            rospy.loginfo("takefoff")
        elif key.code == SPACE:
            self.pub_land.publish(self.empty_msg)
            rospy.loginfo("land")
        elif key.code == BACKSPACE:
            self.pub_reset.publish(self.empty_msg)
            rospy.loginfo("reset")

        elif key.code == UP:
            self.ardrone_vel.linear.z += VELOCITY_INCLIMENT
        elif key.code == DOWN:
            self.ardrone_vel.linear.z -= VELOCITY_INCLIMENT
        elif key.code == LEFT:
			self.ardrone_vel.angular.z += VELOCITY_INCLIMENT
        elif key.code == RIGHT:
			self.ardrone_vel.angular.z -= VELOCITY_INCLIMENT

        elif key.code == KEY_w:
            self.ardrone_vel.linear.x += VELOCITY_INCLIMENT
        elif key.code == KEY_s:
            self.ardrone_vel.linear.x -= VELOCITY_INCLIMENT

        elif key.code == KEY_a:
            self.ardrone_vel.linear.y += VELOCITY_INCLIMENT
        elif key.code == KEY_d:
            self.ardrone_vel.linear.y -= VELOCITY_INCLIMENT
    
    def keyupCallback(self, key):
        pass
    
    def publish_vel(self):
        rate = rospy.Rate(10)
        while not rospy.is_shutdown():
            self.pub_vel.publish(self.ardrone_vel)
            rate.sleep()
    

if __name__ == '__main__':
    operator = KeyboardOperation()
    operator.subscribe_key_data()
    operator.publish_vel()

